docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,202 @@
+objects in node. Thus in order to understand AsyncWrap one must first understand
+handle objects.
+
+The API that node exposes is mostly defined in JavaScript files. However
+ECMAScript does not define any API for creating a TCP socket, reading a file
+etc.. That logic is implemented in C++ using libuv and the v8 API. The JavaScript in
+nodecore interacts with this C++ layer using the handle objects.
+
+For example to create and connect a TCP socket:
+
+```javascript
+const TCP = process.binding('tcp_wrap').TCP;
+const TCPConnectWrap = process.binding('tcp_wrap').TCPConnectWrap;
+
+const req = new TCPConnectWrap();
+req.oncomplete = oncomplete;
+req.address = address;
+req.port = port;
+
+const socket = new TCP();
+socket.onread = onread;
+socket.connect(req, address, port);
+```
+
+This example uses two handle objects. The first one (`TCPConnectWrap`) is for
+connecting the socket, the second one (`TCP`) is for maintaining the connection.
+
+`TCPConnectWrap` gets its information by setting properties on the handle
+object, like `address` and `port`. Those properties are read by the C++ layer,
+but can also be inspected from the AsyncWrap hooks. When the handle is create
+using `new TCPConnectWrap()` the `init` hook is called.
+
+A `oncomplete` property is also set, this is the callback for when the connection is made or failed. Just before calling `oncomplete` the `before` hook
+is called, just after the `after` hook is called.
+
+The `TCP` handle works exactly the same way, except that the information
+is passed as arguments to a method `.connect` and the `onread` function
+is called multiply times, thus it behaves like an event. This also means that
+the `before` and `after` hooks are called multiply times.
+
+Thus one should expect the hooks be called in the following order:
+
+```javascript
+init // TCPConnectWrap
+init // TCP
+=== tick ===
+before // TCPConnectWrap
+after // TCPConnectWrap
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+...
+```
+
+_tick_ indicates there is at least one tick between the text above and the text below.
+
+## The API
+
+At the moment there don't exist a high level API for AsyncWrap. It is simply exposed through `process.binding`:
+
+```javascript
+const asyncWrap = process.binding('async_wrap');
+```
+
+_Be warned that this API is not an official API and can change at any time, even if it's just patch
+update._
+
+To assign the hooks call:
+
+```javascript
+asyncWrap.setupHooks(init, before, after);
+function init(provider) { /* consumer code */ }
+function before() { /* consumer code */ }
+function after() { /* consumer code */ }
+```
+
+Note that calling `asyncWrap.setupHooks` again, will overwrite the previous
+hooks.
+
+#### Enable and disable
+
+Because there is a small performance penalty in just calling a noop function,
+AsyncWrap is not enabled by default. To enable AsyncWrap call:
```

Comment:
> I still think that the call to enable() should be required

Elaborate?

As for making them optional, that's a trivial patch at this point because of one of the most recent patches.